File: SortVision/public/code/selection/rust/selectionSort.rs
/// Find minimum element in unsorted portion
/// # Arguments
/// * `arr` - Vector to search in
/// * `start` - Starting index of unsorted portion
/// * `end` - Ending index of unsorted portion
/// # Returns
/// Index of minimum element

fn find_min_index(arr: &Vec<i32>, start: usize, end: usize) -> usize {
    // Initialize minimum index as the start position
    let mut min_idx = start;
    
    // Iterate through the unsorted portion from start+1 to end
    for i in (start + 1)..=end {
        // If we find a smaller element, update min_idx
        if arr[i] < arr[min_idx] {
            min_idx = i;
        }
    }
    
    // Return the index of the minimum element found
    min_idx
}

/// Public function to sort a vector using Selection Sort algorithm
/// Selection Sort works by repeatedly finding the minimum element 
/// from the unsorted portion and placing it at the beginning.
/// # Arguments
/// * `arr` - Vector to be sorted
pub fn sort(arr: &mut Vec<i32>) {
    let len = arr.len();
    if len <= 1 {
        return; // Already sorted
    }
    
    // Iterate through the array from 0 to second-to-last element
    for i in 0..len-1 {
        // Find the minimum element in the unsorted portion
        let min_idx = find_min_index(arr, i, len - 1);
        
        // Swap the found minimum element with the first element of unsorted portion
        // Only swap if we found a smaller element (when min_idx is different from i)
        if min_idx != i {
            arr.swap(i, min_idx);
        }
    }
}

File: SortVision/public/code/selection/python/selectionSort.py
def find_min_index(arr: list[int], start: int, end: int) -> int:
    """
    Find minimum element in unsorted portion
    Args:
        arr: List to search in
        start: Starting index of unsorted portion
        end: Ending index of unsorted portion
    Returns:
        int: Index of minimum element
    """
    # Initialize the minimum index as the start position
    min_idx = start
    
    # Iterate through the unsorted portion to find the minimum element
    for i in range(start + 1, end):
        # If we find a smaller element, update min_idx
        if arr[i] < arr[min_idx]:
            min_idx = i
    
    return min_idx

def selection_sort(arr: list[int]) -> list[int]:
    
    # Get the length of the array
    n = len(arr)
    
    # If array is empty or has one element, it's already sorted
    if n <= 1:
        return arr
    
    # Iterate through the array
    for i in range(n - 1):
        # Find the minimum element in the unsorted portion
        min_idx = find_min_index(arr, i, n)
        
        # Swap the found minimum element with the first element
        # of the unsorted portion if they are different
        if i != min_idx:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
    
    return arr